design/notifications.md

```diff
@@ -0,0 +1,405 @@
+them because they're recorded using synchronuous HTTP calls. Because of this, the endpoint in the service that provides
+the backwards compatible REST API for notifications will add the timestamp to the notification request before publishing
+it to the exchange. For updates that are posted directly to AMQP, the service itself will be responsible for including
+the timestamp in the request.
+
+#### Structure
+
+The portion of the event recorder service that listens for incoming requests and dispatches them to handlers doesn't
+have to change based on message category, so we can code it in a fairly straight forward manner. The event recorder will
+subscribe to the routing key wildcard pattern, `events.*.update.*`. As mentioned above, the second and fouth components
+of the routing key are the category and the update type. Once a message arrives, it will be dispatched to a handler
+based on the category. If no handler was registered for the category, then the AMQP message is simply acknowledged and
+ignored. If a handler has been registered for the category then the handler is called in order to process the message
+further. The main listener function does very little else at this point. If the handler returns without an error, then
+the message will be acknowleged. If an error that is likely to be recoverable (that is, if reprocessing the message at a
+later time is likely to succeed) is returned, then the message is negatively acknowleged goes back into the queue for
+later processing. At this time, no limits will be placed on message processing reattempts so that we can be sure to
+process each message that has a reasonable chance of succeeding. If an error that is not likely to be recoverable is
+returned then the most likely cause is a bug somewhere in our system, either in the service that sent the initial
+request or in the event recorder itself. For this reason, an email will be sent to a configurable email address whenever
+an unrecoverable error is detected. In CyVerse's DE deployment, we'll use `intercom-de@cyverse.org` so that an Intercom
+issue will be created and we can follow up later. (At some later time, we may want to have alerts like these go to a
+system monitoring tool such as Prometheus instead.)
+
+Unrecoverable errors will be distinguished from recoverable errors by type. Errors of type `UnrecoverableError` will be
+treated as unrecoverable. Errors of any other type will be treated as recoverable. A special `RecoverableError` type
+will also be created to help make it clear when recoverable and unrecoverable errors are being created.
+
+One need that is likely to arise is code reuse. I entertained the idea of using interface embedding along with a
+template method pattern, but after some thought, this approach seems to be unnecessary. Instead, each message handler
+will implement this interface:
+
+``` go
+type MessageHandler interface {
+    HandleMessage(messageBody map[string]interface{}) error
+}
+```
```

Comment:
This is probably an implementation detail note, but I've had good experiences in the search and async-tasks services with creating a sort of "plugin" system -- create a wrapping type, e.g. `AsyncTasksUpdater` for async-tasks, that has a `map[string]BehaviorProcessor` as a field and methods to add things to that map, then expecting the caller to do something like

```go
updater := NewAsyncTasksUpdater(...)
updater.AddBehavior(<key>, <updateFn>)
```

Within each handler, for turning `map[string]interface{}` into real types, github.com/mitchellh/mapstructure has been really nice.

Hopefully these ideas are helpful, and they'd make this consistent with some other code, admittedly code that pretty much only I've touched :sweat_smile: